### ISSUE TITLE
insertMany() error handling middleware does not work

### ISSUE THREAD

**Issue by williamdang** [Automattic/mongoose#6228]
**Do you want to request a *feature* or report a *bug*?**
I want to report a bug

**What is the current behavior?**
The error handling middleware for `insertMany()` does not work.

**If the current behavior is a bug, please provide the steps to reproduce.**
Here is a script to reproduce the problem:

    const Mongoose = require('mongoose');
    const Schema = Mongoose.Schema;
    const expect = require('expect.js');
    
    describe('Mongoose', function () {
      Mongoose.Promise = global.Promise;
      
      let canHookPostInsertMany = false;
      let canHookPostInsertManyError = false;
      
      const fooSchema = new Schema({
        foo: {type: String},
      });
      
      fooSchema.index({foo: 1}, {unique: true});
      
      fooSchema.post('insertMany', (doc, next) => {
        canHookPostInsertMany = true;
        next();
      });
      
      fooSchema.post('insertMany', (err, doc, next) => {
        canHookPostInsertManyError = true;
        next(err);
      });
      
      const Foo = Mongoose.model('Foo', fooSchema);
      
      before(async function () {
        await Mongoose.connect('mongodb://localhost/test');
        
        await Foo.remove().exec();
      });
      
      after(async function () {
        await Foo.remove().exec();
      });
      
      describe('duplicate insertMany()', function () {
        it('should turn on "canHookPostInsertMany" and "canHookPostInsertManyError"', async function () {
          try {
            const foos = await Foo.insertMany([
              {foo: 'duplicated'},
              {foo: 'duplicated'},
            ]);
          } catch (e) {
          }
          
          expect(canHookPostInsertMany).to.be(true);
          expect(canHookPostInsertManyError).to.be(true);
        });
      });
    });

**What is the expected behavior?**
The post "insertMany()" hook for error handling should work. 

**Please mention your node.js, mongoose and MongoDB version.**
I am using node.js 8.9.4, mongoose 5.0.9 and MongoDb 3.6.3

**Comment by lineus** [comment]
Hi @williamdang, Thanks for making such an awesome example! In this case, I think the 2nd [faq](http://mongoosejs.com/docs/faq.html) in the list covers this scenario. unique is not going to
*actually* be unique until the indexes are built in mongodb, there's an example that covers how to wait for the 'index' event covered in the faq as well. Let me know if you have any questions.

**Comment by lineus** [comment]
Hi @williamdang. Thanks for the great repro script!

Apologies for the hasty first comment. I came back to this after a little while because it wasn't sitting right with me.

it looks like the model middleware for insertMany doesn't conform to the error handler (3 arg) middleware mentioned in the docs [here](http://mongoosejs.com/docs/middleware.html#error-handling)

it does however send the first arg as an error. I made a few adjustments to your tests
   1. since it doesn't honor the special 3 arg error handler middleware I removed that post
   hook from your example and just tested the first arg for instance of err in the remaining one.
   1. I broke the two examples out to show that a working example gets past the test for err
   in the post hook and sets the correct variable.
   1. I also tested the caught error for the duplicate error code.

```
const Mongoose = require('mongoose')
const Schema = Mongoose.Schema
const expect = require('chai').expect

describe('Mongoose', function () {
  Mongoose.Promise = global.Promise

  let canHookPostInsertMany = false
  let canHookPostInsertManyError = false

  const fooSchema = new Schema({
    foo: { type: String }
  })

  fooSchema.index({ foo: 1 }, { unique: true })

  fooSchema.post('insertMany', (doc, next) => {
    if (doc instanceof Error) {
      canHookPostInsertManyError = true
      return next(doc)
    }
    canHookPostInsertMany = true
    return next()
  })

  const Foo = Mongoose.model('Foo', fooSchema)

  before(async function () {
    await Mongoose.connect('mongodb://localhost/test')

    await Foo.remove().exec()
  })

  after(async function () {
    await Foo.remove().exec()
  })

  describe('duplicate insertMany()', function () {
    it('should turn on "canHookPostInsertManyError"', async function () {
      try {
        const foos = await Foo.insertMany([
          { foo: 'duplicated' },
          { foo: 'duplicated' }
        ])
      } catch (e) {
        expect(e.message).to.match(/^E11000/)
      }
      expect(canHookPostInsertManyError).to.deep.equal(true)
    })
    it('should turn on "canHookPostInsertMany', async function () {
      await Foo.insertMany([ new Foo({ foo: 'notDuplicated' }) ])
      expect(canHookPostInsertMany).to.deep.equal(true)
    })
  })
})
```
I'll see if I can figure out if the lack of Error Handling Middleware on insertMany is intentional or a bug and update here when I find out. Unless someone confirms one way or the other first :)

**Comment by lineus** [comment]
## To illustrate the difference a little more clearly:
and to save the code that I'm using to step through the debugger

### insertManyTest.js
```
'use strict'

const mongoose = require('../lib/test_db')
const Schema = mongoose.Schema

const fooSchema = new Schema({
  name: String
})

fooSchema.index({ foo: 1 }, { unique: true })

fooSchema.post('insertMany', (err, res, next) => {
  console.log('post.insertMany:error', err.name)
  next()
})

const Test = mongoose.model('tests', fooSchema)

Test.remove({}).then(() => {
  Test.insertMany({ foo: 'duplicated' }).then(() => {
    Test.insertMany({ foo: 'duplicated' }).then(() => {
    }).catch(() => { mongoose.connection.close() })
  })
})
```
### output: 
```
InspiredMacPro:6228 lineus$ node ./insertManyTest.js 
InspiredMacPro:6228 lineus$ 
```
### saveTest.js
```
'use strict'

const mongoose = require('../lib/test_db')
const Schema = mongoose.Schema

const fooSchema = new Schema({
  name: String
})

fooSchema.index({ foo: 1 }, { unique: true })

fooSchema.post('save', (err, res, next) => {
  console.log('post.save:error', err.name)
  next()
})

const Test = mongoose.model('tests', fooSchema)

Test.remove({}).then(() => {
  Test.create({ foo: 'duplicated' }).then(() => {
    Test.create({ foo: 'duplicated' }).then(() => {
    }).catch(() => { mongoose.connection.close() })
  })
})
```
### output:
```
InspiredMacPro:6228 lineus$ node ./saveTest.js 
post.save:error BulkWriteError
InspiredMacPro:6228 lineus$ 
```

I believe that I've isolated the difference between how the save and the insertMany post hooks get called. I'll play around with a solution. @vkarpov15 @varunjayaraman if the difference here is by design, and I shouldn't try to change this behavior, let me know. If that's the case, I'll submit a PR against the docs instead.

**Comment by vkarpov15** [comment]
Thanks for your patience, fixed in master and fix will be in 5.0.12 :+1:
